github/messages.go

```diff
@@ -132,17 +138,40 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
 //       // Process payload...
 //     }
 //
-func ValidatePayload(r *http.Request, secretKey []byte) (payload []byte, err error) {
-	payload, err = ioutil.ReadAll(r.Body)
-	if err != nil {
-		return nil, err
+func ValidatePayload(r *http.Request, secretKey []byte) ([]byte, error) {
+	// jsonPayload is the json webhook data
+	var jsonPayload []byte
+	// rawPayload is the raw body that GitHub uses to calculate the signature
+	var rawPayload []byte
+
+	switch c := r.Header.Get("Content-Type"); c {
+	case "application/x-www-form-urlencoded":
+		// If the content type is application/x-www-form-urlencoded
+		// the json payload will be under the "payload" form param
+		jsonPayload = []byte(r.FormValue(payloadFormParam))
+
+		// GitHub calculates the signature based on the query-escaped
+		// post body. In order to validate, we need to reconstruct the raw body.
+		rawPayload = []byte(fmt.Sprintf("%s=%s", payloadFormParam, url.QueryEscape(string(jsonPayload))))
```

Comment:
It feels to me that using `fmt.Sprintf` here is not being helpful, and the code would be simpler with just plain string concatenation.

How about:

```Go
rawPayload = []byte(payloadFormParam + "=" + url.QueryEscape(string(jsonPayload)))
```